### PR TITLE
[WIP] feat: Try out CircleCI's new infrastructure for test & release flow

### PR DESCRIPTION
**Pull request by mburns**
Based off of @hugomrdias's [gitlab prototyping](https://github.com/hugomrdias/js-ipfs) (along with [jenkins](https://ci.ipfs.team/blue/organizations/jenkins/IPFS%2Fjs-ipfs/detail/master/22/)) and inspired by our internal use of Circle CI.

This prototypes out what a modern CircleCI setup would look like for our testing & release needs.

TODO:
 - [x] artifacts
 - [x] caching
 - [x] mac os x (node{8, 10})
 - [x] linux (node{8, 10})
 - [ ] (pre-)release flow to github+npm
 - [ ] interop with go-ipfs test releases

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+      - checkout
+      - run: npx nyc -s npm run test:node
+      - save_cache:
+          key: ipfs-{{ .BuildNum }}
+          paths:
+            - .nyc_output/
+
+  test-browser-chrome:
+    docker:
+      - image: hugomrdias/node-chrome:test]
```
Comment (alanshaw) [inline .circleci/config.yml]: `hugomrdias`? Also, trailing `]`?

File: .circleci/config.yml [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+    steps:
+      - checkout
+      - run: AEGIR_BROWSERS=ChromeDocker npm run test:browser
+      - store_artifacts:
+          path: coverage/
+
+  test-browser-firefox:
```
Comment (alanshaw) [inline .circleci/config.yml]: oooh new ✨

File: .circleci/config.yml [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+      - run: npm --no-git-tag-version version $bump &>/dev/null
+      - run: npx conventional-changelog -i CHANGELOG.md -r 0 -s -p angular
+      - run: git add CHANGELOG.md
+      - run: version=`node -p "require('./package.json').version"`
+      - run: mv -f _package.json package.json
+      - run: git commit -am "docs(CHANGELOG) $version"
+      - run: npm version $bump -m "chore(release) v%s"
+      - run: git push --follow-tags upstream master
+      - run: npm publish
+      - run: npx gh-release -y
```
Comment (alanshaw) [inline .circleci/config.yml]: Can you explain what's going on for these two jobs: pre-release and release?

File: .circleci/config.yml [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+
+  interop:
+    docker:
+      - image: node:10
+    steps:
+      - run: curl --request POST --form "token=$CI_JOB_TOKEN" --form ref=master https://gitlab.com/api/v4/projects/8952909/trigger/pipeline
```
Comment (alanshaw) [inline .circleci/config.yml]: So this triggers a CI job on Gitlab for interop...it should use the commit that this job is building - looks like it just uses master?

File: .circleci/config.yml [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+
+experimental:
+  notify:
+    branches:
+      only:
+        - master
```
Comment (alanshaw) [inline .circleci/config.yml]: What does this do?

File: .circleci/config.yml [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+
+workflows:
+  version: 2
+  build_all:
+    jobs:
+      - build_linux
+      - build_macos:
+          filters:
+            branches:
+              only: master
```
Comment (alanshaw) [inline .circleci/config.yml]: Only master for macos?

File: .circleci/config.yml (hunk 2 above)
Comment (alanshaw) [reply .circleci/config.yml]: N.B. We need to add testing for webworker here too.

**Comment by alanshaw** [comment]
@mburns looks like there's issues with the config:

```console
2 schema violations found
  required key [jobs] not found
  required key [version] not found
```

```console
Configuration errors: 2 errors occurred:

* Error parsing config file: yaml: line 5: could not find expected ':'
* Cannot find a job named `build` to run in the `jobs:` section of your configuration file.
If you expected a workflow to run, check your config contains a top-level key called 'workflows:'
```

I'd really love to move to circle or gitlab asap. Do you have ay idea when you might be able to look into this? I'd be happy to merge this if we get it working in order to try it out temporarily and leave the following two tasks for a separate PR:

 - (pre-)release flow to github+npm
 - interop with go-ipfs test releases

File: .circleci/config.yml (hunk 1 above)
Comment (hugomrdias) [reply .circleci/config.yml]: no

File: .circleci/config.yml [hunk 7]
```diff
@@ -0,0 +1,154 @@
+version: 2
+jobs:
+  check:
+    docker:
+      - image: node:10
+    steps:
+      - checkout
+      - run: npm run lint
+      - store_artifacts:
+          path: dist/
```
[the remark this answers is not in the document]
Comment (hugomrdias) [reply .circleci/config.yml]: no

File: .circleci/config.yml (hunk 4 above)
Comment (hugomrdias) [reply .circleci/config.yml]: in my config this is meant to be a nightly triggers thats why it runs only master

**Comment by alanshaw** [comment]
@hugomrdias do you have any free time to help out with this?

**Comment by hugomrdias** [comment]
i would love too!! but i still waiting on feedback and machines to finish the gitlab prototype so we can compare both.

Also if i can have access to circle ci i can finish the configuration just like a did for gitlab.

For this we need to setup circle ci on my fork to test if the pre-release and release jobs work like the gitlab version and we need windows machines.

This is really important for us js team so i'm available to do whatever its needed to finish this and move away from jenkins.


I also updated and added docs on the gitlab config
https://github.com/hugomrdias/js-ipfs/blob/master/.gitlab-ci.yml

```md
image: hugomrdias/node-alpine:test
stages:
  - check
  - test
  - cov
  - pre-release
  - release
  - interop

.only_changes: &only_changes
  changes:
    - src/**/*
    - test/**/*
    - package.json

variables:
  DOCKER_DRIVER: overlay2

before_script:
  - npm ci -q

check:
  stage: check
  script:
    - npm run lint
    # - npx dependency-check package.json './test/**/*.js'
    # - npm install --no-lockfile @commitlint/config-conventional @commitlint/cli --save-dev
    # - npx commitlint --extends=@commitlint/config-conventional --from=$CI_COMMIT_BEFORE_SHA --to=$CI_COMMIT_SHA
    - npm run build
  only:
    - branches
  artifacts:
    paths:
      - dist/
  except:
    refs:
      - schedules

test-node-8:
  stage: test
  image: node:8
  script:
    - npx nyc -s npm run test:node
  artifacts:
    paths:
      - .nyc_output/
  except:
    refs:
      - schedules

test-node-10:
  stage: test
  script:
    - npx nyc -s npm run test:node
  artifacts:
    paths:
      - .nyc_output/
  except:
    refs:
      - schedules

test-browser-chrome:
  image: hugomrdias/node-chrome:test
  stage: test
  script:
    - AEGIR_BROWSERS=ChromeDocker npm run test:browser
    # - AEGIR_BROWSERS=ChromeDocker npm run test:webworker
  except:
    refs:
      - schedules

test-browser-firefox:
  image: hugomrdias/node-firefox:test
  stage: test
  script:
    - AEGIR_BROWSERS=FirefoxHeadless npm run test:browser
    # - AEGIR_BROWSERS=ChromeDocker npm run test:webworker
  artifacts:
    paths:
      - coverage/
  except:
    refs:
      - schedules

# Cov job does NOT run tests it only aggregates all the code coverage reports from the tests jobs into a single one.
# it also generates a nice html report and auto deploys it on a gitlab page, so we can check coverage line by line.
cov:
  stage: cov
  image: node:10-alpine
  before_script: []
  variables:
    GIT_STRATEGY: none
  script:
    - cp coverage/coverage-final.json .nyc_output/browser.json
    - npx nyc report --reporter text-summary --reporter html
  dependencies:
    - test-node-10
    - test-node-8
    - test-browser-firefox
  artifacts:
    paths:
      - coverage/
  coverage: '/Lines        : \d+\.\d+/'
  except:
    refs:
      - schedules

.git_auth: &git_auth |
  export GH_RELEASE_GITHUB_API_TOKEN=$GITHUB_TOKEN
  git remote add upstream https://hugomrdias:$GITHUB_TOKEN@github.com/hugomrdias/$CI_PROJECT_NAME.git
  git config --global user.email "hugomrdias@gmail.com"
  git config --global user.name "hugomrdias"
  git checkout $CI_COMMIT_REF_NAME
  git fetch upstream
  git branch -u upstream/$CI_COMMIT_REF_NAME
  echo "//registry.npmjs.org/:_authToken=$NPM_TOKEN" > .npmrc

# Does a pre-release without tags this enables us to have changelog for pre-releases and don't mess up release changlog
pre-release:
  stage: pre-release
  before_script: []
  script:
    - *git_auth
    - npm --no-git-tag-version version prerelease --preid=rc &>/dev/null # bump package.json to x.x.x-rc.x but without tagging
    - npx conventional-changelog-cli -i CHANGELOG.md -r 0 -s -p angular -u # generate changelog
    - git add CHANGELOG.md # add changelog
    - version=`node -p \"require('./package.json').version\"` # save the current version to a env var
    - "git commit -am \"chore(pre-release): ${version}\"" # commit everything with nice msg
    - git push upstream master # push
    - npm publish . --tag next # npm publish with the dist-tag next
  only:
    refs:
      - master
  except:
    refs:
      - schedules
    variables:
      - $CI_COMMIT_MESSAGE =~ /^chore\(pre-release\)/
      - $CI_COMMIT_MESSAGE =~ /^chore\(release\)/

# Does a full release this is meant to be a manual job
# You will notice a bit of a back and forth with the package.json
# This happens because conventional-changelog needs to run with package.json already in the next version
# but we also want to run npm version after conventional-changelog so to not bump twice with npm version
# we play a bit with package.json.
release:
  stage: release
  before_script: []
  script:
    - *git_auth
    - cp package.json _package.json # conventional-changelog needs to know the
    - bump=`npx conventional-recommended-bump -p angular` # find the correct bump according to conventional-changelog-angular
    - npm --no-git-tag-version version $bump &>/dev/null # first npm version without tag
    - npx conventional-changelog -i CHANGELOG.md -r 0 -s -p angular # generate changelog
    - git add CHANGELOG.md # git add
    - version=`node -p \"require('./package.json').version\"` # save the current version to a env var
    - mv -f _package.json package.json # revert package.json to the previous version so we can run npm version and not bump twice
    - "git commit -am \"docs(CHANGELOG): ${version}\"" # commit everything with nice msg
    - "npm version $bump -m \"chore(release): v%s\"" # run the final npm version to bump, commmit, tag
    - git push --follow-tags upstream master # push with tags
    - npm publish # publish to npm
    - npx gh-release -y # do a github release with the same changelog created above
  only:
    refs:
      - master
  except:
    refs:
      - schedules
    variables:
      - $CI_COMMIT_MESSAGE =~ /^chore\(pre-release\)/
      - $CI_COMMIT_MESSAGE =~ /^chore\(release\)/
  when: manual

# This job is a cross repo/project job, it's meant to be triggered by a schedule (nightly).
# This triggers interop repo to run tests with the lastest pre-release from js-ipfs
interop:
  stage: interop
  variables:
    GIT_STRATEGY: none
  before_script: []
  script:
    - curl --request POST --form "token=$CI_JOB_TOKEN" --form ref=master https://gitlab.com/api/v4/projects/8952909/trigger/pipeline
  only:
    - schedules

```

**Pull request closed without merging**